hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCompactionChore.java

```diff
@@ -0,0 +1,253 @@
+          MobConstants.DEFAULT_MOB_MAJOR_COMPACTION_REGION_BATCH_SIZE);
+
+  }
+
+  @VisibleForTesting
+  public MobFileCompactionChore(Configuration conf, int batchSize) {
+    this.conf = conf;
+    this.regionBatchSize = batchSize;
+  }
+
+  @Override
+  protected void chore() {
+
+    boolean reported = false;
+
+    try (Connection conn = ConnectionFactory.createConnection(conf);
+        Admin admin = conn.getAdmin();) {
+
+      TableDescriptors htds = master.getTableDescriptors();
+      Map<String, TableDescriptor> map = htds.getAll();
+      for (TableDescriptor htd : map.values()) {
+        if (!master.getTableStateManager().isTableState(htd.getTableName(),
+          TableState.State.ENABLED)) {
+          LOG.info("Skipping MOB compaction on table {} because it is not ENABLED",
+            htd.getTableName());
+          continue;
+        } else {
+          LOG.info("Starting MOB compaction on table {}, checking {} column families",
+            htd.getTableName(), htd.getColumnFamilyCount());
+        }
+        for (ColumnFamilyDescriptor hcd : htd.getColumnFamilies()) {
+          try {
+            if (hcd.isMobEnabled()) {
+              if (!reported) {
+                master.reportMobCompactionStart(htd.getTableName());
+                reported = true;
+              }
+              LOG.info("Major MOB compacting table={} cf={}", htd.getTableName(),
+                hcd.getNameAsString());
+              if (regionBatchSize == MobConstants.DEFAULT_MOB_MAJOR_COMPACTION_REGION_BATCH_SIZE) {
+                LOG.debug("Table={} cf ={}: batch MOB compaction is disabled, {}=0 -"+
+                  " all regions will be compacted in parallel", htd.getTableName(),
+                  hcd.getNameAsString(), "hbase.mob.compaction.batch.size");
+                admin.majorCompact(htd.getTableName(), hcd.getName());
+              } else {
+                LOG.info("Table={} cf={}: performing MOB major compaction in batches "+
+                    "'hbase.mob.compaction.batch.size'={}", htd.getTableName(),
+                    hcd.getNameAsString(), regionBatchSize);
+                performMajorCompactionInBatches(admin, htd, hcd);
+              }
+            } else {
+              LOG.info("Skipping table={} column family={} because it is not MOB-enabled",
+                htd.getTableName(), hcd.getNameAsString());
+            }
+          } catch (IOException e) {
+            String errMsg = String.format("Failed to compact table=%s cf=%s",
+              htd.getTableName(), hcd.getNameAsString());
+            LOG.error(errMsg, e);
+          } catch (InterruptedException ee) {
+            Thread.currentThread().interrupt();
+            master.reportMobCompactionEnd(htd.getTableName());
+            String warnMsg = String.format("Failed to compact table=%s cf=%s",
+              htd.getTableName(), hcd.getNameAsString());
+            LOG.warn(warnMsg, ee);
```

Comment:
Above.

Reply:
the slf4j Logger methods for messages take vararg to do matching to parameters and also have special handling so that the very last parameter can be an exception.

ref: http://www.slf4j.org/faq.html#paramException